File: modules/agent/src/repositories/project_repository.py
"""Repository layer for project data access."""

from datetime import date
from typing import Optional, Dict, Any, List

from sqlalchemy import select, func
from sqlalchemy.orm import selectinload

from lib.db import async_get_session
from models.Project import Project
from models.Deal import Deal
from models.Lead import Lead
from models.LeadProject import LeadProject
from schemas.project import ProjectCreate, ProjectUpdate

__all__ = ["ProjectCreate", "ProjectUpdate"]


async def find_all_projects(tenant_id: Optional[int]) -> List[Project]:
    """Find all projects for tenant."""
    async with async_get_session() as session:
        stmt = select(Project).order_by(Project.name)
        if tenant_id:
            stmt = stmt.where(Project.tenant_id == tenant_id)
        result = await session.execute(stmt)
        return list(result.scalars().all())


async def find_project_by_id(project_id: int, tenant_id: Optional[int]) -> Optional[Project]:
    """Find project by ID."""
    async with async_get_session() as session:
        stmt = select(Project).where(Project.id == project_id)
        if tenant_id:
            stmt = stmt.where(Project.tenant_id == tenant_id)
        result = await session.execute(stmt)
        return result.scalar_one_or_none()


async def count_project_deals(project_id: int) -> int:
    """Count deals for a project."""
    async with async_get_session() as session:
        result = await session.execute(
            select(func.count(Deal.id)).where(Deal.project_id == project_id)
        )
        return result.scalar() or 0


async def count_project_leads(project_id: int) -> int:
    """Count leads for a project."""
    async with async_get_session() as session:
        result = await session.execute(
            select(func.count(LeadProject.lead_id)).where(LeadProject.project_id == project_id)
        )
        return result.scalar() or 0


async def create_project(
    tenant_id: Optional[int],
    name: str,
    description: Optional[str],
    status: Optional[str],
    current_status_id: Optional[int],
    start_date: Optional[date],
    end_date: Optional[date],
    user_id: Optional[int],
) -> Project:
    """Create a new project."""
    async with async_get_session() as session:
        project = Project(
            tenant_id=tenant_id,
            name=name,
            description=description,
            status=status,
            current_status_id=current_status_id,
            start_date=start_date,
            end_date=end_date,
            created_by=user_id,
            updated_by=user_id,
        )
        session.add(project)
        await session.commit()
        await session.refresh(project)
        return project


async def update_project(
    project_id: int,
    tenant_id: Optional[int],
    data: Dict[str, Any],
    user_id: Optional[int],
) -> Optional[Project]:
    """Update a project. Returns None if not found."""
    async with async_get_session() as session:
        stmt = select(Project).where(Project.id == project_id)
        if tenant_id:
            stmt = stmt.where(Project.tenant_id == tenant_id)
        result = await session.execute(stmt)
        project = result.scalar_one_or_none()

        if not project:
            return None

        if data.get("name") is not None:
            project.name = data["name"]
        if data.get("description") is not None:
            project.description = data["description"]
        if data.get("status") is not None:
            project.status = data["status"]
        if data.get("current_status_id") is not None:
            project.current_status_id = data["current_status_id"]
        if data.get("start_date") is not None:
            project.start_date = data["start_date"]
        if data.get("end_date") is not None:
            project.end_date = data["end_date"]
        project.updated_by = user_id

        await session.commit()
        await session.refresh(project)
        return project


async def delete_project(project_id: int, tenant_id: Optional[int]) -> bool:
    """Delete a project. Returns False if not found."""
    async with async_get_session() as session:
        stmt = select(Project).where(Project.id == project_id)
        if tenant_id:
            stmt = stmt.where(Project.tenant_id == tenant_id)
        result = await session.execute(stmt)
        project = result.scalar_one_or_none()

        if not project:
            return False

        await session.delete(project)
        await session.commit()
        return True


async def find_project_leads(project_id: int, tenant_id: Optional[int]) -> List[Lead]:
    """Find leads for a project."""
    async with async_get_session() as session:
        stmt = (
            select(Lead)
            .options(selectinload(Lead.current_status), selectinload(Lead.account))
            .join(LeadProject, Lead.id == LeadProject.lead_id)
            .where(LeadProject.project_id == project_id)
            .order_by(Lead.created_at.desc())
        )
        if tenant_id:
            stmt = stmt.where(Lead.tenant_id == tenant_id)
        result = await session.execute(stmt)
        return list(result.scalars().all())


async def find_project_deals(project_id: int, tenant_id: Optional[int]) -> List[Deal]:
    """Find deals for a project."""
    async with async_get_session() as session:
        stmt = (
            select(Deal)
            .options(selectinload(Deal.current_status))
            .where(Deal.project_id == project_id)
            .order_by(Deal.created_at.desc())
        )
        if tenant_id:
            stmt = stmt.where(Deal.tenant_id == tenant_id)
        result = await session.execute(stmt)
        return list(result.scalars().all())

File: modules/agent/internal/repositories/project_repository.go
package repositories

import (
	"errors"

	apperrors "agent/internal/errors"
	"agent/internal/lib"
	"agent/internal/models"
	"gorm.io/gorm"
)

// ProjectRepository handles project data access
type ProjectRepository struct {
	db *gorm.DB
}

// ProjectCreateInput represents input for creating a project
type ProjectCreateInput struct {
	TenantID        *int64
	Name            string
	Description     *string
	Status          *string
	CurrentStatusID *int64
	StartDate       *string
	EndDate         *string
	CreatedBy       int64
	UpdatedBy       int64
}

// NewProjectRepository creates a new project repository
func NewProjectRepository(db *gorm.DB) *ProjectRepository {
	return &ProjectRepository{db: db}
}

// FindAll returns all projects for a tenant
func (r *ProjectRepository) FindAll(tenantID *int64) ([]models.Project, error) {
	var projects []models.Project

	query := r.db.Model(&models.Project{}).Order("name ASC")

	if tenantID != nil {
		query = query.Where("tenant_id = ?", *tenantID)
	}

	err := query.Find(&projects).Error
	return projects, err
}

// FindByID returns a project by ID
func (r *ProjectRepository) FindByID(id int64) (*models.Project, error) {
	var project models.Project
	err := r.db.First(&project, id).Error
	if errors.Is(err, gorm.ErrRecordNotFound) {
		return nil, apperrors.ErrNotFound
	}
	if err != nil {
		return nil, err
	}
	return &project, nil
}

// Create creates a new project
func (r *ProjectRepository) Create(input ProjectCreateInput) (*models.Project, error) {
	project := &models.Project{
		TenantID:        input.TenantID,
		Name:            input.Name,
		Description:     input.Description,
		Status:          input.Status,
		CurrentStatusID: input.CurrentStatusID,
		CreatedBy:       input.CreatedBy,
		UpdatedBy:       input.UpdatedBy,
	}

	project.StartDate = lib.ParseDateString(input.StartDate)
	project.EndDate = lib.ParseDateString(input.EndDate)

	if err := r.db.Create(project).Error; err != nil {
		return nil, err
	}
	return project, nil
}

// Update updates a project
func (r *ProjectRepository) Update(id int64, updates map[string]interface{}) error {
	return r.db.Model(&models.Project{}).Where("id = ?", id).Updates(updates).Error
}

// Delete deletes a project
func (r *ProjectRepository) Delete(id int64) error {
	return r.db.Delete(&models.Project{}, id).Error
}

// GetDealsCount returns the count of deals for a project
func (r *ProjectRepository) GetDealsCount(projectID int64) (int64, error) {
	var count int64
	err := r.db.Table(`"Lead_Project"`).
		Joins(`JOIN "Lead" ON "Lead".id = "Lead_Project".lead_id`).
		Where(`"Lead_Project".project_id = ? AND "Lead".lead_type = ?`, projectID, "Deal").
		Count(&count).Error
	return count, err
}

// GetLeadsCount returns the count of leads (non-deals) for a project
func (r *ProjectRepository) GetLeadsCount(projectID int64) (int64, error) {
	var count int64
	err := r.db.Table(`"Lead_Project"`).
		Joins(`JOIN "Lead" ON "Lead".id = "Lead_Project".lead_id`).
		Where(`"Lead_Project".project_id = ? AND "Lead".lead_type != ?`, projectID, "Deal").
		Count(&count).Error
	return count, err
}

// GetProjectLeads returns leads for a project (excluding deals)
func (r *ProjectRepository) GetProjectLeads(projectID int64, tenantID *int64) ([]models.Lead, error) {
	var leads []models.Lead
	query := r.db.Table(`"Lead"`).
		Joins(`JOIN "Lead_Project" ON "Lead".id = "Lead_Project".lead_id`).
		Where(`"Lead_Project".project_id = ? AND "Lead".lead_type != ?`, projectID, "Deal")

	if tenantID != nil {
		query = query.Where(`"Lead".tenant_id = ?`, *tenantID)
	}

	err := query.Find(&leads).Error
	return leads, err
}

// GetProjectDeals returns deals for a project
func (r *ProjectRepository) GetProjectDeals(projectID int64, tenantID *int64) ([]models.Lead, error) {
	var deals []models.Lead
	query := r.db.Table(`"Lead"`).
		Joins(`JOIN "Lead_Project" ON "Lead".id = "Lead_Project".lead_id`).
		Where(`"Lead_Project".project_id = ? AND "Lead".lead_type = ?`, projectID, "Deal")

	if tenantID != nil {
		query = query.Where(`"Lead".tenant_id = ?`, *tenantID)
	}

	err := query.Find(&deals).Error
	return deals, err
}
